plugins/cryptolib/cryptolib.cpp

```diff
@@ -0,0 +1,705 @@
+
+        __lenResult = SHA224_DIGEST_LENGTH;//28 bytes
+        __result = CTXMALLOC(parentCtx, __lenResult);
+
+        if(!SHA224_Final((unsigned char *)__result, &context))
+        {
+            free(__result);
+            throwHashError("OpenSSL ERROR calling SHA224_Final");
+        }
+    }
+    else if (0 == stricmp("SHA256", algorithm))
+    {
+        SHA256_CTX context;
+        if(!SHA256_Init(&context))
+            throwHashError("OpenSSL ERROR calling SHA256_Init");
+
+        if(!SHA256_Update(&context, (unsigned char*)inputData, lenInputData))
+            throwHashError("OpenSSL ERROR calling SHA256_Update");
+
+        __lenResult = SHA256_DIGEST_LENGTH;//32 bytes
+        __result = CTXMALLOC(parentCtx, __lenResult);
+
+        if(!SHA256_Final((unsigned char *)__result, &context))
+        {
+            free(__result);
+            throwHashError("OpenSSL ERROR calling SHA256_Final");
+        }
+    }
+    else if (0 == stricmp("SHA384", algorithm))
+    {
+        SHA512_CTX context;//SHA384 uses the SHA512 context
+        if(!SHA384_Init(&context))
+            throwHashError("OpenSSL ERROR calling SHA384_Init");
+
+        if(!SHA384_Update(&context, (unsigned char*)inputData, lenInputData))
+            throwHashError("OpenSSL ERROR calling SHA384_Update");
+
+        __lenResult = SHA384_DIGEST_LENGTH;//48 bytes
+        __result = CTXMALLOC(parentCtx, __lenResult);
+
+        if(!SHA384_Final((unsigned char *)__result, &context))
+        {
+            free(__result);
+            throwHashError("OpenSSL ERROR calling SHA384_Final");
+        }
+    }
+    else if (0 == stricmp("SHA512", algorithm))
+    {
+        SHA512_CTX context;
+        if(!SHA512_Init(&context))
+            throwHashError("OpenSSL ERROR calling SHA512_Init");
+
+        if(!SHA512_Update(&context, (unsigned char*)inputData, lenInputData))
+            throwHashError("OpenSSL ERROR calling SHA512_Update");
+
+        __lenResult = SHA512_DIGEST_LENGTH;//64 bytes
+        __result = CTXMALLOC(parentCtx, __lenResult);
+
+        if(!SHA512_Final((unsigned char *)__result, &context))
+        {
+            free(__result);
+            throwHashError("OpenSSL ERROR calling SHA512_Final");
+        }
+    }
+    else
+    {
+        throw MakeStringException(-1, "Unsupported hash algorithm '%s' specified", algorithm);
+    }
+}
+
+//-------------------------------------------------------------------------------------------------------------------------------------------
+// PUBLIC KEY SUPPORT
+//-------------------------------------------------------------------------------------------------------------------------------------------
+
+CRYPTOLIB_API void CRYPTOLIB_CALL clInstalledPublicKeyAlgorithms(bool & __isAllResult, size32_t & __lenResult, void * & __result)
+{
+    __isAllResult = false;
+    StringArray algorithms;
+#if OPENSSL_VERSION_NUMBER < 0x10100000L
+    CriticalBlock block(loadingDigestCS);
```

Comment:
This critical section needs to be in a nested block - otherwise it will apply the whole of the rest of the function.  Same goes for the other used.  Alternatively extract it into a shared function...